### PR TITLE
Add Jetty10 and Jetty11 modules

### PR DESCRIPTION
**Pull request by minwoox**
Motivations:
It would be nice if we support Jetty 10 and Jetty 11

Modifications:
- Added Jetty10 and Jetty11 module
  - There're quite a few breaking changes when updating to the version 10.
    - `eventListeners` and `lifeCycleListeners` are both `EventListener`. https://github.com/line/armeria/blob/fbab0e3e5f3ba576a5db9bfc8ca1980d60b41006/jetty/jetty9/src/main/java/com/linecorp/armeria/server/jetty/JettyServiceBuilder.java#L51-L52
    - Setting a request content has changed.
    - Creating HttpFields(Http headers) is changed.
    - `HttpChannel` is created from `ArmeriaHttpConnection`.
  - There aren't many breaking changes between 10 and 11 except that 11 uses Jakarta while 10 uses Javax
  - Both requires slf4j 2.x
    - We cannot exclude slf4j 1.x completely at the moment because the core module added it with the api configuration.
    - Will revisit this later if this becomes a problem.
- Updated Jetty 9 version from 9.4.50.v20221201 to 9.4.51.v20230217
- Added Spring Boot Jetty example

Result:
- You can now use Armeria with Jetty 10 and Jetty 11

### CODE REVIEW

File: jetty/jetty11/src/main/java/com/linecorp/armeria/server/jetty/JettyServiceBuilder.java [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.BlockingTaskExecutor;
+
+/**
+ * Builds a {@link JettyService}. Use {@link JettyService#of(Server)} if you have a configured Jetty
+ * {@link Server} instance.
+ */
+public final class JettyServiceBuilder {
```
Comment (ikhoon) [inline jetty/jetty11/src/main/java/com/linecorp/armeria/server/jetty/JettyServiceBuilder.java:45]: How about adding `AbstractJettyServiceBuilder` to share the common part with `JettyServiceBuilder` in `jetty9`?

File: jetty/jetty11/src/main/java/com/linecorp/armeria/server/jetty/ArmeriaEndPoint.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+import org.eclipse.jetty.io.FillInterest;
+import org.eclipse.jetty.io.WriteFlusher;
+import org.eclipse.jetty.util.Callback;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+final class ArmeriaEndPoint implements EndPoint {
```
Comment (ikhoon) [inline jetty/jetty11/src/main/java/com/linecorp/armeria/server/jetty/ArmeriaEndPoint.java:35]: Should we extract the common part of `ArmeriaEndpoint` between Jetty 9 and Jetty 11 to `AbstractArmeriaEndpoint`? The two classes seem quite similar.

**Review by ikhoon — COMMENTED** [review]
Overall, very nice.

File: jetty/jetty11/src/main/java/com/linecorp/armeria/server/jetty/JettyServiceBuilder.java (hunk 1 above)
Comment (minwoox) [reply jetty/jetty11/src/main/java/com/linecorp/armeria/server/jetty/JettyServiceBuilder.java]: That's a good suggestion. Addressed. 👍

**Review by trustin — COMMENTED** [review]
Just some minor comments only :+1:

File: examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+     * Returns a new {@link JettyService} that redirects the incoming requests to the Jetty instance
+     * provided by Spring Boot.
+     */
+    @Bean
+    public JettyService jettyService(ServletWebServerApplicationContext applicationContext) {
+        final JettyWebServer jettyWebServer = jettyServer(applicationContext);
+        assert jettyWebServer instanceof JettyWebServerWrapper;
+        final JettyWebServerWrapper webServerWrapper = (JettyWebServerWrapper) jettyWebServer;
+        return JettyService.of(jettyWebServer.getServer(), null, false,
+                               webServerWrapper::doStart, webServerWrapper::stop);
```
Comment (ikhoon) [inline examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java:48]: IIUC, this is the same with `JettyService.of(jettyWebServer.getServer(), null, false)`.
Could you explain why we need a separate start and stop tasks apart from `jetty.server.Server.start()` and `jetty.server.Server.stop()`

File: examples/spring-boot-jetty/src/test/resources/application-testbed.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Use-legacy-processing to use this application-testbed.yml file instead of application.yml
+spring:
+  config:
+    use-legacy-processing: true
```
[the remark this answers is not in the document]
Comment (minwoox) [reply examples/spring-boot-jetty/src/test/resources/application-testbed.yml:4]: We can still use `application.yml` file. I believe it's that the way of applyinng the configuration is changed. 
https://spring.io/blog/2020/08/14/config-file-processing-in-spring-boot-2-4
Let me handle this issue separately. https://github.com/line/armeria/issues/4899

File: examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    private static class JettyWebServerWrapper extends JettyWebServer {
+
+        JettyWebServerWrapper(Server server) {
+            super(server);
+        }
+
+        @Override
+        public void start() {
+            // no-op. This will prevent Jetty from starting by the Spring Boot.
```
[the remark this answers is not in the document]
Comment (minwoox) [reply examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java:95]: We are starting the Jetty server in `JettyService` using the `ServerListener`.
https://github.com/line/armeria/blob/main/jetty9/src/main/java/com/linecorp/armeria/server/jetty/JettyService.java#L522
https://github.com/line/armeria/blob/main/jetty9/src/main/java/com/linecorp/armeria/server/jetty/JettyService.java#L215

[the remark this answers is not in the document]
Comment (minwoox) [reply examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java:95]: Armeria server is started by the [ArmeriaServerGracefulShutdownLifecycle](https://github.com/line/armeria/blob/57295d3ed82748dc6949f825910705a6c17d098f/spring/boot3-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java#L47) and Jetty server is started when `ServerListener.serverStarting` is called.

File: jetty/jetty11/src/main/java/com/linecorp/armeria/server/jetty/AbstractJettyServiceBuilder.java [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+import org.eclipse.jetty.server.handler.HandlerWrapper;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * Builds a {@link JettyService}. Use {@link JettyService#of(Server)} if you have a configured Jetty
+ * {@link Server} instance.
```
[the remark this answers is not in the document]
Comment (minwoox) [reply jetty/jetty11/src/main/java/com/linecorp/armeria/server/jetty/AbstractJettyServiceBuilder.java]: Updated. 😊

File: dependencies.toml [hunk 7]
```diff
@@ -50,8 +51,14 @@ javax-inject = "1"
 javax-jsr311 = "1.1.1"
 javax-validation = "2.0.1.Final"
 jctools = "4.0.1"
-jetty93 = '9.3.30.v20211001'
-jetty94 = "9.4.50.v20221201"
+# Find the latest version of the major 10 https://mvnrepository.com/artifact/org.eclipse.jetty/jetty-server
```
[the remark this answers is not in the document]
Comment (minwoox) [reply dependencies.toml]: Didn't know about it. Thanks!

File: examples/spring-boot-jetty/build.gradle [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    implementation project(':core')
+    implementation project(':spring:boot3-starter')
+    implementation project(':jetty11')
+
+    implementation libs.slf4j2.api
+    implementation(libs.spring.boot3.starter.jetty) {
+        exclude group: 'jakarta.websocket', module: 'jakarta.websocket-api'
+        exclude group: 'jakarta.websocket', module: 'jakarta.websocket-client-api'
+        exclude group: 'org.eclipse.jetty.websocket', module: 'websocket-jetty-server'
+        exclude group: 'org.eclipse.jetty.websocket', module: 'websocket-jakarta-server'
```
[the remark this answers is not in the document]
Comment (minwoox) [reply examples/spring-boot-jetty/build.gradle]: Oops, I've probably done something wrong. I didn't have to exclude them. 😓

File: examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (trustin) [reply examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java:95]: Then could you rephrase it? e.g.

```
// This method is intentionally no-op, because ... <detailed context> ....
```

[the remark this answers is not in the document]
Comment (trustin) [reply examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java:95]: Please also fix the grammar of `This will prevent Jetty from starting by the Spring Boot.`

**Review by trustin — APPROVED** [review]
Just some nits 👍

File: examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    public ArmeriaServerConfigurator armeriaServiceInitializer(JettyService jettyService) {
+        return sb -> sb.serviceUnder("/", jettyService);
+    }
+
+    @Configuration(proxyBeanMethods = false)
+    @ConditionalOnClass({ Servlet.class, Server.class, Loader.class, WebAppContext.class })
+    static class EmbeddedJetty {
+
+        @Bean
+        JettyServletWebServerFactory jettyServletWebServerFactory(
```
Comment (jrhee17) [inline examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java:65]: I feel like this example would be much more usable if `JettyServletWebServerFactory#setAutoStart` is added 😅 
Having said this, I think the current approach is the best way for now

**Review by jrhee17 — APPROVED** [review]
Looks good 👍   Thanks @minwoox 🙇 👍 🚀

File: examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java (hunk 3 above)
Comment (ikhoon) [reply examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java:48]: IIUC, `JettyWebServerWrapper.start()` performs but doesn't `JettyWebServerWrapper.getServer().start()` starts the Jetty server?
That means we can just pass the unwrapped `jettyWebServer.getServer()` to start the Jetty server by Armeria?

Comment (ikhoon) [reply examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java:48]: I realized that we need to start `JettyWebServer` instead of `jetty.server.Server`.

Comment (minwoox) [reply examples/spring-boot-jetty/src/main/java/example/springframework/boot/jetty/HelloConfiguration.java]: Let Spring Boot manages the lifecycle and rollback `JettyService` to the original form. Thanks! 😉

**Review by ikhoon — APPROVED** [review]
Nice work, @minwoox! 🙇‍♂️
